mne/viz/epochs.py

```diff
@@ -536,3 +714,123 @@ def plot_epochs_psd(epochs, fmin=0, fmax=np.inf, proj=False, n_fft=256,
+    # do the plotting
+    tick_list = []
+    for ii in range(n_channels):
+        ch_ind = ii + params['ch_start']
+        if ii >= len(lines):
+            break
+        elif ch_ind < len(params['picks']):
+            # scale to fit
+            ch_name = params['ch_names'][ch_ind]
+            tick_list += [ch_name]
+            offset = offsets[ii]
+            end = params['t_start'] + params['duration']
+            this_data = data[ch_ind][params['t_start']:end]
+            start_idx = params['t_start'] / len(params['epochs'].times)
+            labels = params['labels'][start_idx:]
+            params['ax'].set_xticklabels(labels)
+
+            # subtraction here gets corect orientation for flipped ylim
+            ydata = offset - this_data
+            lines[ii].set_ydata(ydata)
+            xdata = params['times'][:params['duration']]
+            lines[ii].set_xdata(xdata)
+            vars(lines[ii])['ch_name'] = ch_name
+            color = params['color'][types[ch_ind]]
+            vars(lines[ii])['def_color'] = params['color'][types[ch_ind]]
+            lines[ii].set_color(color)
+        else:
+            # "remove" lines
+            lines[ii].set_xdata([])
+            lines[ii].set_ydata([])
+    # deal with event lines
+    if params['epoch_times'] is not None:
+        # find events in the time window
+        epoch_times = params['epoch_times']
+        event_line = params['event_line']
+        xs = list()
+        ys = list()
+        for tt in epoch_times:
+            xs += [tt, tt, np.nan]
+            ylim = params['ax'].get_ylim()
+            ys += [ylim[0], ylim[1], np.nan]
+        event_line.set_xdata(xs)
+        event_line.set_ydata(ys)
+
+    # finalize plot
+    params['ax'].set_xlim(params['times'][0],
+                          params['times'][0] + params['duration'], False)
+    params['ax'].set_yticklabels(tick_list)
+    params['vsel_patch'].set_y(params['ch_start'])
+    params['fig'].canvas.draw()
+
+
+def _plot_onscroll(event, params):
+    """
+    Scroll events.
+    """
+    orig_start = params['ch_start']
+    if event.step < 0:
+        params['ch_start'] = min(params['ch_start'] + params['n_channels'],
+                                 len(params['ch_names']) -
+                                 params['n_channels'])
+    else:  # event.key == 'up':
+        params['ch_start'] = max(params['ch_start'] - params['n_channels'], 0)
+    if orig_start != params['ch_start']:
+        _channels_changed(params)
+
+
+def _plot_window(value, params):
+    """
+    Deal with horizontal epoch window.
+    """
+    max_times = len(params['times']) - params['duration']
+    if value > max_times:
+        value = len(params['times']) - params['duration']
+    if value < 0:
+        value = 0
+    if params['t_start'] != value:
+        params['t_start'] = value
+        params['hsel_patch'].set_x(value)
+        _plot_traces(params)
+
+
+def _mouse_click(event, params):
```

Comment:
yeah eventually we would want to deduplicate, but for now I'd say get all the functionality in and working well. Then we can figure out how to make it DRY